robottelo/host_helpers/api_factory.py

```diff
@@ -711,6 +713,217 @@ def wait_for_errata_applicability_task(
+                entities['LifecycleEnvironment'] not in entities['ContentView'].environment,
+            ]
+        ):
+            # promote newest version by id
+            entities['ContentView'].version.sort(key=lambda version: version.id)
+            entities['ContentView'].version[-1].promote(
+                data={'environment_ids': entities['LifecycleEnvironment'].id}
+            )
+            read_entities()
+
+        if (  # assign env to ak if not present
+            entities['ActivationKey'].environment is None
+            or entities['ActivationKey'].environment.id != entities['LifecycleEnvironment'].id
+        ):
+            entities['ActivationKey'].environment = entities['LifecycleEnvironment']
+            entities['ActivationKey'].update(['environment'])
+            read_entities()
+        if (  # assign cv to ak if not present
+            entities['ActivationKey'].content_view is None
+            or entities['ActivationKey'].content_view.id != entities['ContentView'].id
+        ):
+            entities['ActivationKey'].content_view = entities['ContentView']
+            entities['ActivationKey'].update(['content_view'])
+
+        # register with now setup entities, using ak
+        read_entities()
+        result = client.register(
+            activation_keys=entities['ActivationKey'].name,
+            target=self._satellite,
+            org=entities['Organization'],
+            loc=None,
+            force=force,
+        )
+        if result.status != 0:
+            pytest.fail(f'Failed to register the host: {client.hostname}.\n{result.stderr}')
+        if not client.subscribed:
+            pytest.fail(
+                f"Failed to subscribe the host: {client.hostname}, to content-view: {entities['ContentView'].name}"
+            )
+        if rex_key:
+            client.add_rex_key(self._satellite)
+
+        # attempt to enable all repositories available to client,
+        # ie any repos added to content-view prior to calling this method.
+        # note: this will fail if no repos are available to client from CV/AK
+        if enable_repos:
+            output = client.execute(r'subscription-manager repos --enable \*')
+            if output.status != 0:
+                pytest.fail(
+                    'Failed to enable all available repositories using subscription-manager.'
+                    f' For client: {client.hostname}.\n{output.stdout}'
+                )
+
+        # For in-between parameterized sessions:
+        #   unregister the host if it's still subscribed.
+        request = MagicMock()
+
+        @request.addfinalizer
+        def cleanup():
+            nonlocal client
+            if client.subscribed:
+                client.unregister()
+            assert (
+                client.subscribed is False
+            ), f'Failed to fully teardown client {client.hostname}, maintains some content association.'
```

Comment:
What is the purpose of this and why is mock being used here?

Reply:
mock is being used to fake a `request` object that we can usually import into fixture only.

The finalizer is added to unregister the host during teardown- 
so you don't have to unregister, or force registration in the test, 
- if the same `rhel_contenthost` is used in a parameterized test session, it will often fail, for already being registered in the previous parameterized run.

If unittest mock is not preferred here, do you think something like this would make more sense?

```
yield {return dict w/ entities}
if client.subscribed: client.unregister()
```

Reply:
This teardown will ~hopefully execute in a similar way to existing teardown for local UI fixture, `registered_contenthost`: https://github.com/SatelliteQE/robottelo/blob/c3e329e9ff808245d91686e9c73a4f775f321618/tests/foreman/ui/test_errata.py#L299

Reply:
Remember my earlier advice that, since this isn't a fixture or a test, you should try to avoid pytest-specific behavior. If you really think that this should also handle teardown, then it sounds more like this not being a helper function and instead being a (or multiple) fixture(s).
If you want to keep it a helper function, then it's best to avoid trying to interact in a pytest-like way.

Reply:
@JacobCallahan I think I will implement this setup method to be used in local UI fixture `registered_contenthost`, 
which could perform the pytest related functions like unregistering the host, as it does already.

Reply:
@damoore044 cool. ping me in here once that change is in